portal/common/lib/rpc_selector.ts

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+    GetDynamicFieldObjectParams,
+    GetObjectParams,
+    MultiGetObjectsParams,
+    SuiClient,
+    SuiObjectResponse,
+} from "@mysten/sui/client";
+import { TESTNET_RPC_LIST } from "./constants";
+
+interface RPCSelectorInterface {
+    getObject(input: GetObjectParams): Promise<SuiObjectResponse>;
+    multiGetObjects(input: MultiGetObjectsParams): Promise<SuiObjectResponse[]>;
+    getDynamicFieldObject(
+        input: GetDynamicFieldObjectParams,
+    ): Promise<SuiObjectResponse>;
+    call<T>(method: string, args: any[]): Promise<T>;
+}
+
+class RPCSelector implements RPCSelectorInterface {
+    private static instance: RPCSelector;
+    private clients: SuiClient[];
+    private selectedClient: SuiClient | undefined;
+
+    private constructor(rpcURLs: string[]) {
+        // Initialize clients.
+        this.clients = rpcURLs.map((rpcUrl) => new SuiClient({ url: rpcUrl }));
+        this.selectedClient = undefined;
+    }
+
+    // Get the singleton instance.
+    public static getInstance(): RPCSelector {
+    if (!RPCSelector.instance) {
+        RPCSelector.instance = new RPCSelector(TESTNET_RPC_LIST);
+    }
+    return RPCSelector.instance;
+    }
+
+    // General method to call clients and return the first successful response.
+    private async invokeWithFailover<T>(methodName: string, args: any[]): Promise<T> {
+        if (this.clients.length === 0) {
+            throw new Error("No available clients to handle the request.");
+        }
+
+        const isNoSelectedClient = !this.selectedClient;
+        if (isNoSelectedClient) {
+            return await this.callFallbackClients<T>(methodName, args);
+        }
```

Comment:
nit: this check could be removed i think, since the following try block would result in `callFallbackClients` if there is no selected client

Reply:
Since the first use of the fallback function is considered default behavior, I’d suggest keeping the if statement, as throwing an exception would likely be more costly for the standard flow.